cryptography/src/bls12381/primitives/poly.rs

```diff
@@ -208,61 +208,63 @@ impl<C: Element> Poly<C> {
     }
 
     /// Recover the polynomial's constant term given at least `t` polynomial evaluations.
-    pub fn recover(t: u32, mut evals: Vec<Eval<C>>) -> Result<C, Error> {
+    ///
+    /// # Warning
+    ///
+    /// This function assumes that each evaluation is unique.
+    pub fn recover<'a, I>(t: u32, evals: I) -> Result<C, Error>
+    where
+        C: 'a,
+        I: IntoIterator<Item = &'a Eval<C>>,
+    {
         // Reference: https://github.com/celo-org/celo-threshold-bls-rs/blob/a714310be76620e10e8797d6637df64011926430/crates/threshold-bls/src/poly.rs#L131-L165
 
-        // Ensure there are enough shares
+        // Ensure we have enough shares
         let t = t as usize;
+        let mut evals = evals.into_iter().collect::<Vec<_>>();
         if evals.len() < t {
-            return Err(Error::InvalidRecovery);
+            return Err(Error::NotEnoughPartialSignatures(t, evals.len()));
         }
 
         // Convert the first `t` sorted shares into scalars
-        let mut err = None;
-        evals.sort_by(|a, b| a.index.cmp(&b.index));
+        evals.sort_by_key(|e| e.index);
         let xs = evals
             .into_iter()
             .take(t)
-            .fold(BTreeMap::new(), |mut m, sh| {
+            .fold(Vec::with_capacity(t), |mut m, sh| {
                 let mut xi = Scalar::zero();
                 xi.set_int(sh.index + 1);
-                if m.insert(sh.index, (xi, sh.value)).is_some() {
-                    err = Some(Error::DuplicateEval);
-                }
+                m.push((sh.index, (xi, &sh.value)));
                 m
             });
-        if let Some(e) = err {
-            return Err(e);
-        }
 
         // Iterate over all indices and for each multiply the lagrange basis
         // with the value of the share
-        let mut acc = C::zero();
-        for (i, xi) in &xs {
-            let mut yi = xi.1.clone();
-            let mut num = Scalar::one();
-            let mut den = Scalar::one();
-
-            for (j, xj) in &xs {
-                if i == j {
-                    continue;
-                }
-
-                // xj - 0
-                num.mul(&xj.0);
-
-                // 1 / (xj - xi)
-                let mut tmp = xj.0;
-                tmp.sub(&xi.0);
-                den.mul(&tmp);
-            }
+        xs.iter().try_fold(C::zero(), |mut acc, (i, (xi, yi))| {
+            let (mut num, den) = xs.iter().fold(
+                (Scalar::one(), Scalar::one()),
+                |(mut num, mut den), (j, (xj, _))| {
+                    if i != j {
+                        // (xj - 0)
+                        num.mul(xj);
+
+                        // 1 / (xj - xi)
+                        let mut tmp = *xj;
+                        tmp.sub(xi);
+                        den.mul(&tmp);
+                    }
+                    (num, den)
+                },
```

Comment:
instead of:

```
if !a {
    // do a bunch of stuff
}
return result
```

consider
```
if a {
    return result
}
// a bunch of stuff that is now less nested by one level
```

this mirrors more closely the control flow you had before when you did:
```
if i == j {
    continue;
}
```

Reply:
The difference earlier is that we aren't required to return `(num, den)` in the `continue` case. I echo your preference but find this is better for `fold`.